documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,586 @@
+
+This is still relatively complex, however, because the worker Docker containers
+must be launched from Airflow, which is itself dockerized. This requires either
+using a Docker-in-Docker approach, which should typically be avoided, or
+modifying the configuration to allow Airflow to launch sibling containers[^1].
+Requirements differ across operating systems, yet this **must** be tested and
+maintained across platforms because the data refresh is used as a critical step
+in setting up our local development environment.
+
+#### Support for progress tracking
+
+With the EC2 approach it is trivial to calculate the `percentage_complete` for
+the reindexing task, and report to Airflow via the task status endpoint.
+Airflow's ECS Operators do not have out-of-the-box support for streaming the
+Cloudwatch logs from the task while it is running, so we would have less insight
+into the status of the task from Airflow.
+
+#### Impact on Deployments
+
+This is the greatest strength of the ECS approach. Each time an ECS task is spun
+up on Fargate, it will pull the Docker image configured in the task definition.
+By using the `latest` tag in this configuration we can ensure that the latest
+indexer-worker image is pulled each time, which means that any changes to the
+reindexing code will become live in production as soon as the new docker image
+is published after merging on `main`.
+
+For the EC2 approach, any changes to the reindexing script or its dependencies
+would require deploying the catalog (and by extension the indexer workers).
+However:
+
+- No deployment would be required for changes to the vast majority of the data
+  refresh implementation, including the data copy, filtered index creation,
+  alias management, and all aspects of orchestrating the indexer workers.
+- _If_ deploying the indexer-worker becomes a frequent issue in the future, we
+  could iterate on the data refresh DAG to automatically update the workers. For
+  example we could use an Airflow
+  [SSHOperator](https://docs.aws.amazon.com/mwaa/latest/userguide/samples-ssh.html)
+  to connect to the instances and pull the latest image before triggering the
+  task. I think this is unlikely enough to be a problem that this does not need
+  to be a requirement of the initial implementation.
+
+### Conclusion
+
+The ECS approach allows us to remove 100% of the server management code, and
+eliminates the need for deployments in essentially every case (unless the task
+definition itself changes). However it is more expensive, more difficult to
+implement, and requires a special secondary implementation for running locally
+that may be prone to cross-platform issues.
+
+The EC2 approach on the other hand is cheaper, quicker to implement, and gets us
+the vast majority of what we want in terms of removing complexity and reducing
+the need for deployments. Consequently I argue here that the EC2 approach is the
+one we should pursue.
```

Comment:
I think the inability to realistically test the data refresh locally, needing an _entirely separate process_ locally, is pretty significant.

> seemingly-unnecessary webserver aspects

In what sense are they unnecessary? ECS still runs a server, it eliminates a small amount of deployment complexity, but it isn't complexity we don't have elsewhere and can solve to make as or almost as simple as our ECS deployments (which would be good for making Airflow, Kibana, Elasticsearch, etc all simpler and easier to understand and maintain).

The difference in long-term cost is significant, with a bit of up-front infrastructure work that will build off of and benefit the work we've done for other services. Why avoid EC2 if we can and regardless _need_ to make it simple to use for the sake of our other services? We have a low level of understanding of EC2, but we need to improve that, and as we have done, everything about the process improves and eliminates the relative complexity of it compared to ECS.

Reply:
> seemingly-unnecessary webserver aspects

The indexer workers run their own webserver in order to accept tasks and issue progress on them. That code could be completely obviated if the containers which were to run on ECS received the range they needed to index from Airflow and then ran directly on that chunk. The containers could spin up, use the range supplied via CLI to run, and then complete and exit. This would remove the need for maintaining _any_ webserver aspects for the indexer workers, since the containers would only be running as long as it takes to index everything.

To be clear, I'm comfortable with the current approach proposed here nonetheless.